// @opsnexus/ui-kit/src/index.ts
// Shared components
export { SeverityBadge } from './components/SeverityBadge';
export { StatusTag } from './components/StatusTag';
export { RootCauseBadge } from './components/RootCauseBadge';
export { AssetGradeTag } from './components/AssetGradeTag';
export { TimeAgo } from './components/TimeAgo';
export { MetricCard } from './components/MetricCard';
export { AIAnalysisPanel } from './components/AIAnalysisPanel';
export { SearchBar } from './components/SearchBar';
export { TimeRangePicker } from './components/TimeRangePicker';
export { FilterBar } from './components/FilterBar';

// 虚拟滚动表格组件
export { VirtualTable } from './components/VirtualTable';
export type { VirtualTableProps, ColumnDef } from './components/VirtualTable';

// 骨架屏组件
export { PageSkeleton } from './components/PageSkeleton';
export { CardSkeleton } from './components/CardSkeleton';
export { ListSkeleton } from './components/ListSkeleton';

// Hooks
export { useColumnConfig } from './hooks/useColumnConfig';
export { useSubAppTheme } from './hooks/useSubAppTheme';

// Shared types
export type { Severity, RootCauseCategory, AssetGrade } from './types';
export type { PageParams, ApiResponse, PageResult } from './types';

// Animation & visualization components
export { ParticleCanvas } from './components/ParticleCanvas';
export { FlipCard } from './components/FlipCard';
export type { FlipCardProps } from './components/FlipCard';
export { MetricFlipCard } from './components/MetricFlipCard';
export { ScanLine } from './components/ScanLine';
export { NumberFlip } from './components/NumberFlip';
export { AITypewriter } from './components/AITypewriter';
export { SparkLine } from './components/SparkLine';
export { RingGauge } from './components/RingGauge';
export { NoiseFunnel } from './components/NoiseFunnel';
export { LiveDot } from './components/LiveDot';
export { ChartCard } from './components/ChartCard';
export { HealthMatrix } from './components/HealthMatrix';
export type { HealthCell } from './components/HealthMatrix';

// Theme
export { SEVERITY_COLORS, ROOT_CAUSE_COLORS, ASSET_GRADE_COLORS } from './theme';
